### PR TITLE
[adam] Add not torch.jit.is_scripting() as a requirement for switching to fused

### PR DESCRIPTION
**Pull request by janeyx99**
A "fix" following https://github.com/pytorch/pytorch/pull/90865. Realized that fused is not compatible with torch.jit.is_scripting() when looking at a later line.

Took the opportunity to make the code cleaner/slightly more performant (with the extends) as well.

### CODE REVIEW

**Review by albanD — APPROVED** [review]

File: torch/optim/adam.py [hunk 1]
```diff
@@ -316,13 +316,16 @@ def adam(params: List[Tensor],
...
-            fused = True
-        else:
-            fused = False
+        all_tensors = []
+        all_tensors.extend(params)
+        all_tensors.extend(grads)
+        all_tensors.extend(exp_avgs)
+        all_tensors.extend(exp_avg_sqs)
+        all_tensors.extend(max_exp_avg_sqs)
+        all_tensors.extend(state_steps)
```
Comment (Skylion007) [inline torch/optim/adam.py:325]: Couldn't you refactor into a nice single itertools.chain, that way you don't need any temporary list at all.

Comment (janeyx99) [reply torch/optim/adam.py:325]: I started with that, but it does not play well with JIT bindings: https://github.com/pytorch/pytorch/pull/91896#discussion_r1069858978

Comment (Skylion007) [reply torch/optim/adam.py:325]: I see, which is annoying because jit bindings shouldn't be necessary since this path isn't part possible under JIT.

Comment (kit1980) [reply torch/optim/adam.py:325]: The seven lines can be changed into just 1 line, no?
`all_tensors = [*params, *grads, *exp_avgs, *exp_avg_sqs, *max_exp_avg_sqs, *state_steps]`

**Pull request closed without merging**